src/jsonrpc.rs

```diff
@@ -1,61 +1,93 @@
 use jsonrpc_derive::rpc;
 use jsonrpc_stdio_server::{jsonrpc_core::*, ServerBuilder};
-use std::sync::{Mutex, mpsc};
+use std::sync::{mpsc, Arc, Mutex};
 use tracing::{info, trace};
 
+use crate::hookfs::HookFs;
+
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum Comm {
-    Ok = 0,
-    Shutdown = 1,
+    Shutdown = 0,
 }
 
-pub async fn start_server(status: anyhow::Result<()>, tx: mpsc::Sender<Comm>) {
+pub async fn start_server(config: RpcImpl) {
     info!("Starting jsonrpc server");
-    let server = new_server(status, tx);
+    let server = new_server(config);
     let server = server.build();
     server.await;
 }
 
-pub fn new_server(status: anyhow::Result<()>, tx: mpsc::Sender<Comm>) -> ServerBuilder {
+pub fn new_server(config: RpcImpl) -> ServerBuilder {
     info!("Creating jsonrpc server");
-    let io = new_handler(status, tx);
+    let io = new_handler(config);
     ServerBuilder::new(io)
 }
 
-pub fn new_handler(status: anyhow::Result<()>, tx: mpsc::Sender<Comm>) -> IoHandler {
+pub fn new_handler(config: RpcImpl) -> IoHandler {
     info!("Creating jsonrpc handler");
     let mut io = IoHandler::new();
-    io.extend_with(RpcImpl { status, tx:Mutex::new(tx) }.to_delegate());
+    io.extend_with(config.to_delegate());
     io
 }
 
 #[rpc]
 pub trait Rpc {
     #[rpc(name = "get_status")]
     fn get_status(&self) -> Result<String>;
+    #[rpc(name = "update")]
+    fn update(&self, config: Vec<InjectorConfig>) -> Result<String>;
 }
 
 pub struct RpcImpl {
-    status: anyhow::Result<()>,
+    status: Mutex<anyhow::Result<()>>,
     tx: Mutex<mpsc::Sender<Comm>>,
+    hookfs: Option<Arc<HookFs>>,
+}
+
+impl RpcImpl {
+    pub fn new(
+        status: Mutex<anyhow::Result<()>>,
+        tx: Mutex<mpsc::Sender<Comm>>,
+        hookfs: Option<Arc<HookFs>>,
+    ) -> Self {
+        Self { status, tx, hookfs }
+    }
 }
 
 impl Drop for RpcImpl {
     fn drop(&mut self) {
-        println!("> Dropping RpcImpl");
+        trace!("Dropping jrpc handler");
     }
 }
 
 impl Rpc for RpcImpl {
     fn get_status(&self) -> Result<String> {
         trace!("rpc get_status called");
-        match &self.status {
+        match &*self.status.lock().unwrap() {
             Ok(_) => Ok("ok".to_string()),
             Err(e) => {
                 let tx = &self.tx.lock().unwrap();
-                tx.send(Comm::Shutdown).expect("Send through channel failed");
+                tx.send(Comm::Shutdown)
+                    .expect("Send through channel failed");
                 Ok(e.to_string())
-            },
+            }
+        }
+    }
+    fn update(&self, config: Vec<InjectorConfig>) -> Result<String> {
+        trace!("rpc update called");
+        if let Err(e) = &*self.status.lock().unwrap() {
+            return Ok(e.to_string());
+        }
+        let injectors = MultiInjector::build(config);
+        if let Err(e) = &injectors {
+            return Ok(e.to_string());
         }
+        futures::executor::block_on((async || {
```

Comment:
Maybe we need async json-rpc like [jsonrpc-v2](https://github.com/kardeiz/jsonrpc-v2).

Reply:
There can be poly-fill like `async_trait` and the lib itself seems to run on some multithread environment. I will look into it.

Reply:
Oops, async trait is already being used in toda. That's awsome.

Reply:
https://github.com/paritytech/jsonrpc/issues/571
It seems that they are still on the way. And I see no urgent need on using an async structure at once. Maybe we can put it aside and wait for the issue.